### DISCUSSION TITLE
How to lazyload pdf in ngx-extended-pdf-viewer

### DISCUSSION THREAD

**Discussion by subhasispanda98** [stephanrauh/ngx-extended-pdf-viewer#981] [Q&A]
i am fetching pdf from server like 100mb to 500mb and its taking a lot of time to load how can i show some of the pdf like 5 page first then on scroll to bottom load next 5 pages which should fetch from server please help how to do this ?

**Comment by stephanrauh** [comment]
Activate range requests on the server. You don't have to implement fancy logic to preload pages. pdf.js supports this out-of-the-box. Usually, it's a missing server configuration.

https://pdfviewer.net/extended-pdf-viewer/range-requests

**Reply by stephanrauh** [reply]
I've seen a follow-up question in my inbox, but for some reason I don't see it here. I'm pretty sure nginx supports range request out-of-the-box. Please set up an nginx server and deploy your application on it (including the large PDF file). You'll probably notice a big difference to the server you're currently using. Among other things, you can see the range requests in the network tab of your browser.

If you don't want to set up an nginx on your computer, use a docker image. That's easier, anyways.

Now that you know that range request work and that speed up your application, you can analyze why your server doesn't support range requests.

**Reply by subhasispanda98** [reply]
thanks for your valuable information

**Comment by ibtissemoussaifi** [comment]
Hello,
Can you help me please?
I am fetching pdf from Amazon s3 cloudFront and i want to activate range requests
Regards.

**Reply by stephanrauh** [reply]
Amazon claims it works out-of-the-box: https://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/RangeGETs.html

**Reply by ibtissemoussaifi** [reply]
Ok thanks for that.

**Reply by ibtissemoussaifi** [reply]
You mean that  i dont need to configure  Aws S3 to use Range request because it works out-of-the-box, but i have the same problem i can't activate the range request,
I dont know what is the problem, please help !!

**Reply by stephanrauh** [reply]
You've mentioned Cloudfront. Just to make sure I get you correctly: you're accessing the PDF file via a standard http:// oder https:// address?

**Reply by ibtissemoussaifi** [reply]
I accessed the pdf via https

**Reply by ibtissemoussaifi** [reply]
Even if I fetch a pdf (http or https) from google for example (https://docs.aws.amazon.com/general/latest/gr/aws-general.pdf) and not from aws S3 the range request does not activate

**Reply by stephanrauh** [reply]
Strange. I'll try to have a look, but I'm afraid I'm busy this week-end, so I can't promise anything.

**Reply by ibtissemoussaifi** [reply]
ok thanks  :)

**Reply by stephanrauh** [reply]
Range request work nicely on my website: https://www.beyondjava.net/presentations/CDK.pdf 

<img width="1727" alt="image" src="https://user-images.githubusercontent.com/3045767/182019098-4436bf86-fda8-47b4-8311-9abf8280c82d.png">

I didn't use any fancy configuration. Just just ngx-extended-pdf-viewer 14.0.7, plus Cloudfront backed by S3:
```html
<ngx-extended-pdf-viewer [src]="src" [useBrowserLocale]="true"
[showPresentationModeButton]="true"
backgroundColor="black"
[showOpenFileButton]="false"
[showRotateButton]="false"
[showSecondaryToolbarButton]="false"
[pageViewMode]="'single'"
[zoom]="'auto'"
> </ngx-extended-pdf-viewer>
```

```ts
@Component({
  selector: 'app-presentations',
  templateUrl: './presentations.component.html',
  styleUrls: ['./presentations.component.css'],
})
export class PresentationsComponent implements OnInit {
  public src = '/blog/presentations/GraalVM.pdf';
}
```

```ts
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { CoreModule } from '../core/core.module';
import { Routes, RouterModule } from '@angular/router';
import { NgxExtendedPdfViewerModule } from 'ngx-extended-pdf-viewer';
import { PresentationsComponent } from './presentations.component';

const routes: Routes = [
  {
    path: ':talk',
    component: PresentationsComponent,
  },
];

@NgModule({
  declarations: [PresentationsComponent],
  imports: [CommonModule, CoreModule, RouterModule.forChild(routes), NgxExtendedPdfViewerModule],
})
export class PresentationsModule {}
```

**Reply by stephanrauh** [reply]
Please check if your browser also requests range request on https://www.beyondjava.net/presentations/CDK.pdf, just in case.

**Reply by stephanrauh** [reply]
If you're still stuck, please add the attribute `[minifiedJSLibraries]="false"`. Now you can debug the core library. Open the file pdf-*.js, find the implementation of validateRangeRequestCapabilities, and add a breakpoint.

https://github.com/stephanrauh/pdf.js/blob/cf777662071af514c9f2f26c2c94fe63ee66d3e4/src/display/network_utils.js#L24-L71

**Reply by ibtissemoussaifi** [reply]
Hello,
Sorry for the late reply ,
Thank you so much for your answer, it works for me :)
